week5/Q1.py

```diff
@@ -0,0 +1,149 @@
+class OrientedGraph:
+	'''
+	Implementation of oriented graph. All of the nodes have unique values.
+
+	Fields:
+		nodes: set of Nodes, all nodes of the graph
+	'''
+	def __init__(self):
+		self.nodes = set()
+
+	def add(self, value, parent_value = None):
+		'''
+		Adds new Node to the OrientedGraph or updates existing ones.
+
+		Args:
+			value: str
+			parent_value: str
+		'''
+		node = None
+		parent = None
+		for inode in self.nodes:
+			if inode.value == value:
+				node = inode
+			if inode.value == parent_value:
+				parent = inode
+
+		if parent_value:
+			if node:
+				node.parents = node.parents.union({parent})
+				parent.children = parent.children.union({node})
+			else:
+				node = Node(value, parents = {parent})
+				parent.children = parent.children.union({node})
+				self.nodes.add(node)
+		else:
+			if not node:
+				node = Node(value)
+				self.nodes.add(node)
+
+
+class Node:
+	'''
+	Element of OrientedGraph.
+
+	Fields:
+		value: str
+		parents: set of Nodes, from which Node has incoming edges
+		children: set of Nodes, to which Node has outcoming edges
+	'''
+	def __init__(self, value = None, parents = set(), children = set()):
+		self.value = value
+		self.parents = parents
+		self.children = children
```

Comment:
Optional: you could extract the graph data structure into a separate module (i.e., file). So that you could later use it again in some other code (and write tests for the graph separately).